src/Application.vala

```diff
@@ -230,32 +231,43 @@ public class Terminal.Application : Gtk.Application {
         var window = (MainWindow) active_window;
         bool new_window;
 
-        if (window == null || options.lookup ("new-window", "b", out new_window) && new_window) {
+        if (window == null || (options.lookup ("new-window", "b", out new_window) && new_window)) {
             /* Uncertain whether tabs should be restored when app is launched with working directory from commandline.
              * Currently they are set to restore (subject to the restore-tabs setting).
              * If it is desired that tabs should never be restored in these circimstances add another check below.
              */
-            bool restore_tabs = !("commandline" in options || "execute" in options) || window == null;
-            window = new MainWindow (this, restore_tabs);
+            bool restore_tabs = !("commandline" in options || "execute" in options || "working-directory" in options) || window == null;
+            window = new MainWindow (this, restore_tabs, false);
         }
 
-        unowned var working_directory = command_line.get_cwd ();
         unowned string[] commands;
         unowned string command;
+        unowned string wd;
         bool new_tab, minimized;
 
         options.lookup ("new-tab", "b", out new_tab);
+        var wd_option_present = options.lookup ("working-directory", "^&ay", out wd);
 
         if (options.lookup ("execute", "^a&ay", out commands)) {
             for (var i = 0; commands[i] != null; i++) {
                 if (commands[i] != "\0") {
-                    window.add_tab_with_working_directory (working_directory, commands[i], new_tab);
+                    window.add_tab_with_working_directory (
+                        wd != null ? wd : command_line.get_cwd (),
+                        commands[i],
+                        new_tab
+                    );
                 }
             }
         } else if (options.lookup ("commandline", "^&ay", out command) && command != "\0") {
-            window.add_tab_with_working_directory (working_directory, command, new_tab);
-        } else {
-            window.add_tab_with_working_directory (working_directory, null, new_tab);
+            window.add_tab_with_working_directory (
+                wd != null ? wd : command_line.get_cwd (),
+                command,
+                new_tab
+            );
+        } else if (wd_option_present) {
+            window.add_tab_with_working_directory (wd != null ? wd : command_line.get_cwd (), null, new_tab);
+        } else if (new_tab) {
+            window.add_tab_with_working_directory (Environment.get_current_dir (), null, new_tab);
```

Comment:
`Environment.get_current_dir()` here (and amost everywhere in the code) is wrong.

the only places we can be sure it's correct is in `main()`, `new Application()`, `construct`, `local_command_line()`, and `handle_local_options()`.